File: scraper/lib/notify.js
/**
 * Per-run email notification.
 *
 * WHY IT LIVES IN THE SCRAPER, NOT THE BACKEND
 * The backend has no mail code, no mail dependency and no mail settings, so
 * "POST the summary to the backend" would mean writing a brand-new provider
 * integration in Python plus an authenticated endpoint plus a deploy — and it
 * would still call the same provider HTTP API in the end. Worse, it breaks the
 * one requirement that matters: a FAILED run must be able to report itself. The
 * scraper's most likely failures are exactly the ones where the backend is
 * implicated (scraper-config unreachable, scraper-complete rejected), so routing
 * the alert through the backend would silence the alarm precisely when it is
 * needed.
 *
 * Resend's REST API is called with Node's global `fetch` — no SDK, so the
 * dependency list stays at playwright + dotenv + @supabase/supabase-js.
 *
 * FAILURE POLICY: sending mail must never change the outcome of a run. A send
 * failure is logged as an ERROR and swallowed; it does not mask a successful
 * scrape, and it does not rescue a failed one.
 */

const { redact } = require("./logger");

const RESEND_ENDPOINT = "https://api.resend.com/emails";

/** Extract the address from a `Name <addr@host>` form, else return as-is. */
function unwrapAddress(token) {
  const m = String(token).match(/<([^>]+)>/);
  return m ? m[1].trim() : String(token).trim();
}

/**
 * Split a recipient list. Commas and semicolons separate; whitespace does too,
 * EXCEPT inside a `Name <addr>` form, which is split off first so a display name
 * is not shredded into bogus tokens.
 */
function splitRecipients(raw) {
  const angled = [];
  const rest = String(raw).replace(/[^,;]*<[^>]+>/g, (m) => {
    angled.push(unwrapAddress(m));
    return "";
  });
  return [...angled, ...rest.split(/[,;\s]+/)].map((s) => s.trim()).filter(Boolean);
}

/**
 * Deliberately loose: enough to catch a stray word or a missing @, without
 * pretending to implement RFC 5322. Bounded so one very long token cannot make the
 * match pathological.
 */
const isAddress = (s) => s.length <= 320 && /^[^@\s]+@[^@\s]+\.[^@\s]+$/.test(s);

/**
 * Split one env string into usable and unusable recipients in a single pass.
 * These were two functions with the same predicate NEGATED, which had to stay in
 * sync or a token would end up neither used nor warned about.
 */
function classifyRecipients(raw) {
  const valid = [];
  const invalid = [];
  for (const token of raw ? splitRecipients(raw) : []) {
    (isAddress(token) ? valid : invalid).push(token);
  }
  return { valid, invalid };
}

function parseRecipients(raw) { return classifyRecipients(raw).valid; }

/** Recipients present in the env but not shaped like addresses — worth warning about. */
function invalidRecipients(raw) { return classifyRecipients(raw).invalid; }

function readConfig(env = process.env) {
  const { valid, invalid } = classifyRecipients(env.SCRAPER_NOTIFY_EMAILS || "");
  return {
    recipients: valid,
    invalid,
    apiKey: env.RESEND_API_KEY || "",
    from: env.SCRAPER_NOTIFY_FROM || "",
    // "always" (default) or "failure" — some operators only want to hear about problems.
    when: String(env.SCRAPER_NOTIFY_ON || "always").toLowerCase(),
  };
}

/** One-line verdict for the subject line. */
function overallStatus(summary) {
  if (summary.fatalError) return "FATAL";
  const failed = summary.schools.filter((s) => s.status !== "ok").length;
  const partial = summary.schools.filter(
    (s) => s.status === "ok" && ((s.uploadErrors || 0) > 0
      || (s.incompleteExports || []).length > 0
      || (s.exportFailures || []).length > 0),
  ).length;
  if (failed > 0) return "FAILED";
  if (partial > 0) return "PARTIAL";
  if (summary.schools.every((s) => (s.uploaded || 0) === 0)) return "NO-DATA";
  return "OK";
}

function buildSubject(summary) {
  const status = overallStatus(summary);
  const scope = summary.schoolFilter ? summary.schoolFilter : "all schools";
  const uploaded = summary.schools.reduce((n, s) => n + (s.uploaded || 0), 0);
  return `[GAINS scrape] ${status} — ${scope} — ${uploaded} file(s)`
    + `${summary.dryRun ? " (dry-run)" : ""}`;
}

/**
 * Plain-text body. Deliberately text-first: it is what lands in a phone
 * notification preview, it survives every client, and it is greppable.
 * The body leads with the verdict and the reasons — an operator should not have
 * to scroll to learn whether they need to act.
 */
function buildBody(summary) {
  const L = [];
  const status = overallStatus(summary);

  L.push(`Status:    ${status}`);
  L.push(`Run id:    ${summary.runId}`);
  L.push(`Scope:     ${summary.schoolFilter || "all active schools"}${summary.assessmentFilter ? ` | assessment~"${summary.assessmentFilter}"` : ""}`);
  L.push(`Started:   ${summary.startedAt}`);
  L.push(`Duration:  ${summary.durationSec}s`);
  L.push(`Mode:      ${summary.dryRun ? "dry-run (no export/upload)" : "live"}`);
  L.push(`Env:       backend=${summary.backendBaseUrl} storage=${summary.supabaseHost} bucket=${summary.bucket}`);
  L.push("");

  if (summary.fatalError) {
    L.push("FATAL — the run aborted before completing:");
    L.push(`  ${summary.fatalError.name}: ${summary.fatalError.message}`);
    L.push("");
  }

  for (const s of summary.schools) {
    const errs = s.uploadErrors || 0;
    const inc = s.incompleteExports || [];
    const expf = s.exportFailures || [];
    L.push(`── ${s.school} [${s.status.toUpperCase()}]`);
    L.push(`     assessments discovered: ${s.discovered ?? "?"}`);
    L.push(`     exported: ${s.exported ?? s.discovered ?? "?"} | skipped: ${s.skipped ?? 0} | files uploaded: ${s.uploaded ?? 0}`);
    if (s.status !== "ok") {
      L.push(`     ERROR [${s.errorClass}] ${s.error}`);
    }
    if (errs) L.push(`     upload errors: ${errs}`);
    for (const e of expf) L.push(`     export FAILED: ${e.id} "${e.title}" — ${e.reason}`);
    for (const e of inc) L.push(`     incomplete triplet: ${e.id} "${e.title}" missing ${e.missing.join(", ")}`);
    for (const w of s.keyWarnings || []) L.push(`     key sanitised: ${w}`);
    // The funnel is the whole point of the zero-discovery case: without it the
    // email says "0 assessments" and the operator still has to open the log.
    if ((s.discovered === 0 || s.status !== "ok") && s.funnel) {
      const f = s.funnel;
      L.push(`     why nothing was discovered: courses=${f.courses} sections=${f.sections} `
        + `assignmentsSeen=${f.assignments}`);
      const reasons = [
        ["unmapped subject/grade (course skipped)", f.coursesSkippedSubjectGrade],
        ["section had no matching grading category", f.sectionsSkippedNoMatchingCategory],
        ["section had zero grading categories", f.sectionsWithZeroCategories],
        ["not type=assessment", f.notAssessmentType],
        ["grading category did not match the regex", f.categoryUnmatched],
        ["excluded by --assessment", f.filteredByAssessmentFlag],
        ["outside the due window", f.outsideDueWindow],
        ["no due date (use --include-undated)", f.undatedExcluded],
        ["due in the future (use --due-until)", f.futureDated],
        ["unreadable web_url", f.badWebUrl],
      ].filter(([, n]) => n > 0);
      for (const [why, n] of reasons) L.push(`       ${n} × ${why}`);
    }
    L.push("");
  }

  if (summary.storageKeys && summary.storageKeys.length) {
    L.push(`Uploaded objects (${summary.storageKeys.length}):`);
    for (const k of summary.storageKeys.slice(0, 40)) L.push(`  ${k}`);
    if (summary.storageKeys.length > 40) {
      L.push(`  … and ${summary.storageKeys.length - 40} more`);
    }
    L.push("");
  }

  L.push("Next step: the backend worker lands these into raw_* and stops.");
  L.push("It does NOT rebuild reports — see docs/audit/fixes/03_prod_data_sync_runbook.md.");
  L.push("");
  L.push("Check the ledger:");
  L.push("  SELECT run_id, status, files_processed, rows_inserted, error_count");
  L.push("  FROM ingestion_runs ORDER BY started_at DESC LIMIT 3;");

  return L.join("\n");
}

/**
 * Send the run-summary email. Never throws.
 * @returns {Promise<{sent: boolean, reason?: string, recipients?: number}>}
 */
async function sendRunSummary(summary, { env = process.env, log = null, fetchImpl = null } = {}) {
  const cfg = readConfig(env);
  const doFetch = fetchImpl || globalThis.fetch;

  if (cfg.invalid.length) {
    log?.warn("ignoring malformed entries in SCRAPER_NOTIFY_EMAILS", {
      ignored: cfg.invalid.join(","),
    });
  }

  if (!cfg.recipients.length) {
    // Not an error: notification is opt-in. Say so once so an operator who
    // EXPECTED mail can see why none arrived.
    log?.info("no run-summary email sent: SCRAPER_NOTIFY_EMAILS is unset or has no valid addresses");
    return { sent: false, reason: "no recipients" };
  }

  // Rendering is inside a try because it walks caller-supplied summary data.
  // It used to run before the try, so a malformed ledger entry threw OUT of a
  // function documented as never throwing — and since main() awaits this from a
  // `finally`, that rejection replaced the run's real outcome AND lost the email.
  let status;
  try {
    status = overallStatus(summary);
  } catch (err) {
    log?.error("could not derive run status for the notification email", err);
    status = "UNKNOWN";
  }
  if (cfg.when === "failure" && (status === "OK" || status === "NO-DATA")) {
    log?.info("run-summary email suppressed by SCRAPER_NOTIFY_ON=failure", { status });
    return { sent: false, reason: "suppressed by policy" };
  }

  if (!cfg.apiKey || !cfg.from) {
    // Recipients were configured but the transport was not — that IS a
    // misconfiguration and must be loud, or alerts silently never arrive.
    log?.error(
      "run-summary email NOT sent: recipients are configured but "
      + `${!cfg.apiKey ? "RESEND_API_KEY" : "SCRAPER_NOTIFY_FROM"} is missing`,
      null,
      { recipients: cfg.recipients.length },
    );
    return { sent: false, reason: "transport not configured" };
  }

  if (typeof doFetch !== "function") {
    log?.error("run-summary email NOT sent: global fetch unavailable (needs Node >= 18)");
    return { sent: false, reason: "no fetch" };
  }

  let payload;
  try {
    payload = {
      from: cfg.from,
      to: cfg.recipients,
      // Redacted like a log line: the body embeds third-party error strings, and
      // email is a wider egress channel than stdout — it must not be the weaker
      // hygiene path for the same text.
      subject: redact(buildSubject(summary)),
      text: redact(buildBody(summary)),
    };
  } catch (err) {
    // A summary we cannot render must still produce an alert — a bare "the run
    // finished and the report broke" beats silence.
    log?.error("could not render the run-summary email; sending a minimal fallback", err);
    payload = {
      from: cfg.from,
      to: cfg.recipients,
      subject: `[GAINS scrape] ${status} — summary could not be rendered`,
      text: `Run ${summary?.runId || "(unknown)"} finished with status ${status}, but the `
        + `summary could not be rendered: ${err && err.message}\n\n`
        + "Check the run log directly.",
    };
  }

  try {
    const res = await doFetch(RESEND_ENDPOINT, {
      method: "POST",
      headers: {
        Authorization: `Bearer ${cfg.apiKey}`,
        "Content-Type": "application/json",
      },
      body: JSON.stringify(payload),
      // A hung mail provider must not hold a finished scrape open.
      signal: AbortSignal.timeout(15000),
    });
    if (!res.ok) {
      const body = await res.text().catch(() => "");
      log?.error("run-summary email rejected by provider", null, {
        status: res.status, body: body.substring(0, 200),
      });
      return { sent: false, reason: `provider ${res.status}` };
    }
    log?.info("run-summary email sent", { recipients: cfg.recipients.length, status });
    return { sent: true, recipients: cfg.recipients.length };
  } catch (err) {
    // Swallowed by design: a mail outage must not change the run's verdict.
    log?.error("run-summary email failed to send", err, { recipients: cfg.recipients.length });
    return { sent: false, reason: err.name || "send failed" };
  }
}

module.exports = {
  parseRecipients,
  invalidRecipients,
  readConfig,
  overallStatus,
  buildSubject,
  buildBody,
  sendRunSummary,
  RESEND_ENDPOINT,
};

File: backup/lib/notify.py
"""Per-run email notification for the backup service.

Mirrors `scraper/lib/notify.js`'s design decisions — not its code: the never-
throws contract on `send_run_summary`, the recipient-parsing/classification
rules, the failure taxonomy (no-recipients is INFO, everything past that is
ERROR-and-swallowed), the "always" vs. "failure" notify policy, and the
verdict-first subject line. The summary shape is this service's own (dump/
upload figures, not schools/assessment-funnel data) — see
`docs/backup-service-technical-plan.md` Section 1.

WHY A RESEND OUTAGE MUST NEVER SURFACE HERE: `backup.py` calls
`send_run_summary()` from its own `finally` block so the email fires on every
exit path, including ones where an exception is already in flight. If this
function raised, it would replace the run's real outcome with a notification
failure and the operator would never learn the backup itself succeeded or
failed.
"""

import logging
import re
from typing import Any

import httpx

from lib.config import get_settings
from lib.logger import redact

RESEND_ENDPOINT = "https://api.resend.com/emails"

_ADDRESS_RE = re.compile(r"^[^@\s]+@[^@\s]+\.[^@\s]+$")
_ANGLE_ADDR_RE = re.compile(r"[^,;]*<([^>]+)>")

_CLEANUP_LABEL = "Cleanup:            "


def split_recipients(raw: str) -> list[str]:
    """Split a recipient list. Commas/semicolons/whitespace separate, except
    inside a `Name <addr>` form, which is split off first so a display name
    is not shredded into bogus tokens."""
    angled: list[str] = []

    def _capture(match: re.Match[str]) -> str:
        angled.append(match.group(1).strip())
        return ""

    rest = _ANGLE_ADDR_RE.sub(_capture, raw)
    tokens = angled + re.split(r"[,;\s]+", rest)
    return [t.strip() for t in tokens if t.strip()]


def is_address(value: str) -> bool:
    # Deliberately loose: enough to catch a stray word or a missing @, without
    # implementing RFC 5322. Bounded so one very long token can't be pathological.
    return len(value) <= 320 and bool(_ADDRESS_RE.match(value))


def classify_recipients(raw: str) -> tuple[list[str], list[str]]:
    valid: list[str] = []
    invalid: list[str] = []
    for token in split_recipients(raw) if raw else []:
        (valid if is_address(token) else invalid).append(token)
    return valid, invalid


def overall_status(summary: dict[str, Any]) -> str:
    fatal_error = summary.get("fatal_error")
    if not fatal_error:
        return "OK"
    # A tagged pg_dump/upload failure is an expected, named failure mode of the
    # run itself (FAILED); anything else — preflight, or a truly unhandled
    # exception with no stage — means the run aborted somewhere the plan didn't
    # account for (FATAL).
    stage = fatal_error.get("stage") if isinstance(fatal_error, dict) else None
    if stage in ("pg_dump", "upload"):
        return "FAILED"
    return "FATAL"


def _database_host(summary: dict[str, Any]) -> str:
    return str(summary.get("database_host") or "?")


def _human_bytes(num_bytes: Any) -> str:
    if not isinstance(num_bytes, (int, float)) or isinstance(num_bytes, bool):
        return "n/a"
    size = float(num_bytes)
    for unit in ("B", "KB", "MB", "GB"):
        if size < 1024 or unit == "GB":
            return f"{int(size)} {unit}" if unit == "B" else f"{size:.1f} {unit}"
        size /= 1024
    return f"{size:.1f} TB"


def _format_duration(value: Any) -> str:
    if isinstance(value, float):
        return f"{value:.1f}"
    return str(value if value is not None else 0)


def _cleanup_line(cleanup: dict[str, Any] | None, retention_count: Any) -> str:
    if cleanup is None:
        return f"{_CLEANUP_LABEL}not attempted"
    list_error = cleanup.get("list_error")
    if list_error:
        return f"{_CLEANUP_LABEL}FAILED to list old backups: {list_error}"
    errors = cleanup.get("errors") or []
    deleted = cleanup.get("deleted") or []
    remaining = cleanup.get("remaining")
    if errors:
        keys = ", ".join(str(e.get("key")) for e in errors)
        line = f"{_CLEANUP_LABEL}FAILED to delete {len(errors)} old backup(s): {keys}"
        if deleted:
            line += f" (deleted {len(deleted)} other(s) successfully)"
        if remaining is not None:
            line += (
                f" — {remaining} backup(s) now in R2 "
                f"(retry needed to reach the {retention_count}-backup limit)"
            )
        return line
    if deleted:
        return (
            f"{_CLEANUP_LABEL}deleted {len(deleted)} old backup(s) to stay within the "
            f"{retention_count}-backup limit — {remaining} backup(s) now in R2"
        )
    return (
        f"{_CLEANUP_LABEL}nothing to delete — {remaining} of {retention_count} "
        "backup(s) kept (within limit)"
    )


def build_subject(summary: dict[str, Any]) -> str:
    status = overall_status(summary)
    dump_bytes_human = _human_bytes(summary.get("dump_bytes"))
    duration = _format_duration(summary.get("duration_sec"))
    suffix = " (dry-run)" if summary.get("dry_run") else ""
    return f"[GAINS backup] {status} — {dump_bytes_human} — {duration}s{suffix}"


def build_body(summary: dict[str, Any]) -> str:
    status = overall_status(summary)
    lines = [
        f"Status:      {status}",
        f"Run id:      {summary.get('run_id', '?')}",
        f"Started:     {summary.get('started_at', '?')}",
        f"Duration:    {_format_duration(summary.get('duration_sec'))}s",
        f"Mode:        {'dry-run (no dump/upload)' if summary.get('dry_run') else 'live'}",
        f"Database:    {_database_host(summary)}",
        f"Retention:   keep last {summary.get('retention_count', '?')} backup(s)",
        "",
    ]

    fatal_error = summary.get("fatal_error")
    if fatal_error:
        lines.append("FATAL — the run aborted before completing:")
        name = fatal_error.get("name", "?") if isinstance(fatal_error, dict) else "?"
        message = (
            fatal_error.get("message", "?") if isinstance(fatal_error, dict) else str(fatal_error)
        )
        lines.append(f"  {name}: {message}")
        stderr_tail = fatal_error.get("stderr_tail") if isinstance(fatal_error, dict) else None
        if stderr_tail:
            lines.append("")
            lines.append("  pg_dump stderr (tail):")
            lines.append(str(stderr_tail))
        lines.append(f"{_CLEANUP_LABEL}skipped (backup did not complete successfully)")
        lines.append("")
    elif summary.get("dry_run"):
        lines.append(f"{_CLEANUP_LABEL}skipped (dry-run)")
        lines.append("")

    if not summary.get("dry_run") and not fatal_error:
        lines.append(f"Dump size:          {_human_bytes(summary.get('dump_bytes'))}")
        lines.append(f"Dump duration:      {_format_duration(summary.get('dump_duration_sec'))}s")
        lines.append(f"Upload duration:    {_format_duration(summary.get('upload_duration_sec'))}s")
        lines.append(f"R2 bucket:          {summary.get('r2_bucket', '?')}")
        lines.append(f"R2 key:             {summary.get('r2_key', '?')}")
        lines.append(_cleanup_line(summary.get("cleanup"), summary.get("retention_count", "?")))
        lines.append("")

    return "\n".join(lines)


def send_run_summary(summary: dict[str, Any], log: logging.LoggerAdapter) -> dict[str, Any]:  # type: ignore[type-arg]
    try:
        settings = get_settings()
    except Exception as exc:
        log.error(
            "run-summary email NOT sent: configuration failed to load",
            extra={"error": redact(str(exc))},
        )
        return {"sent": False, "reason": "config error"}

    valid, invalid = classify_recipients(settings.backup_notify_emails)
    if invalid:
        log.warning(
            "ignoring malformed entries in BACKUP_NOTIFY_EMAILS",
            extra={"ignored": ",".join(invalid)},
        )

    if not valid:
        # Not an error: notification is opt-in. Say so once so an operator who
        # EXPECTED mail can see why none arrived.
        log.info(
            "no run-summary email sent: BACKUP_NOTIFY_EMAILS is unset or has no valid addresses"
        )
        return {"sent": False, "reason": "no recipients"}

    # Deriving the status walks caller-supplied summary data, so it runs inside
    # its own try — a malformed summary must still produce SOME notification
    # attempt, not an exception out of a function documented as never throwing.
    try:
        status = overall_status(summary)
    except Exception as exc:
        log.error(
            "could not derive run status for the notification email",
            extra={"error": redact(str(exc))},
        )
        status = "UNKNOWN"

    notify_on = (settings.backup_notify_on or "always").strip().lower()
    if notify_on == "failure" and status == "OK":
        log.info(
            "run-summary email suppressed by BACKUP_NOTIFY_ON=failure", extra={"status": status}
        )
        return {"sent": False, "reason": "suppressed by policy"}

    if not settings.resend_api_key or not settings.backup_notify_from:
        # Recipients were configured but the transport was not — that IS a
        # misconfiguration and must be loud, or alerts silently never arrive.
        missing = "RESEND_API_KEY" if not settings.resend_api_key else "BACKUP_NOTIFY_FROM"
        log.error(
            f"run-summary email NOT sent: recipients are configured but {missing} is missing",
            extra={"recipients": len(valid)},
        )
        return {"sent": False, "reason": "transport not configured"}

    try:
        subject = redact(build_subject(summary))
        text = redact(build_body(summary))
    except Exception as exc:
        # A summary that can't be rendered must still produce an alert — a bare
        # "the run finished with status X and the report broke" beats silence.
        log.error(
            "could not render the run-summary email; sending a minimal fallback",
            extra={"error": redact(str(exc))},
        )
        subject = f"[GAINS backup] {status} — summary could not be rendered"
        text = (
            f"Run {summary.get('run_id', '(unknown)')} finished with status {status}, but the "
            f"summary could not be rendered: {exc}\n\nCheck the run log directly."
        )

    payload = {"from": settings.backup_notify_from, "to": valid, "subject": subject, "text": text}

    try:
        response = httpx.post(
            RESEND_ENDPOINT,
            headers={
                "Authorization": f"Bearer {settings.resend_api_key}",
                "Content-Type": "application/json",
            },
            json=payload,
            timeout=15.0,
        )
    except Exception as exc:
        # Swallowed by design: a mail outage must not change the run's verdict.
        log.error(
            "run-summary email failed to send",
            extra={"error": redact(str(exc)), "recipients": len(valid)},
        )
        return {"sent": False, "reason": type(exc).__name__ or "send failed"}

    if response.status_code < 200 or response.status_code >= 300:
        log.error(
            "run-summary email rejected by provider",
            extra={"status": response.status_code, "body": redact(response.text[:200])},
        )
        return {"sent": False, "reason": f"provider {response.status_code}"}

    log.info("run-summary email sent", extra={"recipients": len(valid), "status": status})
    return {"sent": True, "reason": None}
